### PR TITLE
Search API - Refactoring of config template importer 

### PR DESCRIPTION
**Pull request by mtodor**
Refactoring in order to support Search API configuration templates.

Features:
- support for config per field (for example Facets configuration)
- introduce configuration template collection

### CODE REVIEW

File: src/ConfigTemplateDefinition.php [hunk 1]
```diff
@@ -100,17 +112,21 @@ public static function createFromFile($file_name) {
    * @param string $key
    *   The key for template.
    *
-   * @return array|string
+   * @return mixed
    *   Returns value of base template definition property.
    */
   public function getKey($key) {
+    if (!isset($this->definition[$key])) {
+      return NULL;
```
Comment (dbosen) [inline src/ConfigTemplateDefinition.php:103]: not all places, where this is called are handling the NULL correctly. e.g in class ConfigTemplateDefinitionCollection it is assumed, that this is always an array. also a couple of times in class ConfigTemplateDefinitionResolver.

Comment (dbosen) [reply src/ConfigTemplateDefinition.php:103]: Also, is it correct, that this can return string or array? This is a bit strange.

File: src/ConfigTemplateDefinition.php [hunk 2]
```diff
@@ -100,17 +112,21 @@ public static function createFromFile($file_name) {
...
+    }
+
     return $this->definition[$key];
   }
 
   /**
    * Get template source configuration.
    *
-   * @return array
+   * @return array|bool
```
Comment (dbosen) [inline src/ConfigTemplateDefinition.php:112]: This possible bool return value is also not checked in places where getSourceConfig() is called

File: src/ConfigTemplateDefinitionCollection.php [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+   *   Returns TRUE if collection is valid, otherwise it throws.
+   *
+   * @throws \Exception
+   */
+  protected function validate(): bool {
+    foreach ($this->templates as $template_name => $template) {
+      /** @var \Drupal\testsite_builder\ConfigTemplateDefinition $config_template_definition */
+      $config_template_definition = $this->getDefinition($template_name);
+      if ($config_template_definition->getSourceConfig() === FALSE) {
+        throw new \Exception("Unable to find configuration {$template['source']} used by template collection {$this->id}.");
```
Comment (dbosen) [inline src/ConfigTemplateDefinitionCollection.php:88]: This fails with strange messages, when source is either not set, or set as an array. This is a result of the parts in `ConfigTemplateDefinition::getKey()`that I commented on.

File: src/ConfigTemplateDefinitionResolver.php [hunk 4]
```diff
@@ -0,0 +1,403 @@
...
+   *   The field manager service.
+   */
+  public function __construct(ConfigTemplateDefinitionCollection $collection, ConfigTemplateTypePluginManager $config_template_type_manager, EntityTypeManagerInterface $entity_type_manager, EntityFieldManagerInterface $field_manager) {
+    $this->configTemplateTypeManager = $config_template_type_manager;
+    $this->entityTypeManager = $entity_type_manager;
+    $this->fieldManager = $field_manager;
+
+    $this->collection = $collection;
+    $this->entityType = $collection->getEntityType();
+  }
```
Comment (dbosen) [inline src/ConfigTemplateDefinitionResolver.php:86]: Removal of the create function was not what I meant in my previous comment. The create function should stay, it is the stable interface for child classes. Just move the ``` $this->collection = $collection;
    $this->entityType = $collection->getEntityType();```to the constructor, as you did here. Add the `$collection` to the `$instance = new static()`part and just return the `$instance`
In the future, if we decide to change the constructor, we would keep the ::create() interface stable and child classes who do not touch the constructor would be fine.

File: src/ConfigTemplateDefinitionResolver.php [hunk 5]
```diff
@@ -302,30 +274,28 @@ protected function createAllFieldConfigurations(string $entity_type, string $bun
...
 
       /** @var \Drupal\testsite_builder\ConfigTemplateMerge $config_template_merge */
-      $config_template_merge = $config_template_type_plugin->getConfigChangesForField($this->collection->getId(), $entity_type, $bundle, $field_name, $field_config);
+      $config_template_merge = $config_template_type_plugin->getConfigChangesForField($this->collection->getId(), $this->entityType, $bundle, $field_name, $field_config);
       $field_config = $config_template_merge->applyMerge($field_config, []);
     }
 
-    $config_name = ConfigName::createByFullName(implode('.', $config_parts));
+    $source_config_name = ConfigName::createByFullName($this->templateDefinition->getKey('source'));
+    $config_name = ConfigName::createByTypeName($source_config_name->getType(), $config_id);
```
Comment (dbosen) [inline src/ConfigTemplateDefinitionResolver.php:322]: I like the solution, to completely get rid if $config_parts!

File: src/ConfigTemplateDefinition.php [hunk 6]
```diff
@@ -23,11 +30,11 @@ class ConfigTemplateDefinition {
...
 
   /**
-   * The config reverter service from config update.
+   * The directory where template is located.
    *
-   * @var \Drupal\config_update\ConfigRevertInterface
+   * @var string
    */
-  protected $configReverter;
+  protected $templateDirectory;
```
Comment (dbosen) [inline src/ConfigTemplateDefinition.php]: This is never set and used, as well as the corresponding `getTemplateDirectory()`method.

File: src/ConfigTemplateImporter.php [hunk 7]
```diff
@@ -49,194 +42,33 @@ class ConfigTemplateImporter {
...
-    $config = $this->applyConfigurationForBundle($bundle, $config, $dynamic_view_definitions);
-
-    // Copy post generate.
-    foreach ($this->templateDefinition->getKey('post_generate_clone') as $copy_config_key) {
-      $config = $this->copyFromSource($config, $copy_config_key['from'], $copy_config_key['to']);
-    }
-
-    // Create ID for config.
-    $source_config = $this->templateDefinition->getSourceConfig();
-    $config['id'] = "{$source_config['id']}_{$bundle}";
```
Comment (dbosen) [inline src/ConfigTemplateImporter.php:150]: Just out of curiosity: Why did you change the implicit creation of the id into an explicit declaration in the config template?

File: src/ConfigTemplateTypeInterface.php [hunk 8]
```diff
@@ -25,11 +27,15 @@ interface ConfigTemplateTypeInterface extends PluginInspectionInterface {
    * @return \Drupal\testsite_builder\ConfigTemplateMerge
    *   Returns configuration template merge for config type and new field name.
    */
-  public function getConfigChangesForField(string $entity_type, string $bundle, string $field_name, $source_field_config);
+  public function getConfigChangesForField(string $collection_id, string $entity_type, string $bundle, string $field_name, $source_field_config);
```
Comment (dbosen) [inline src/ConfigTemplateTypeInterface.php:30]: The collection_id is introduced for just one plugin and creates much noise in the PR, since it gets shuffled around all the plugins, that totally ignore it.
I don't have a really good solution, but maybe we can still use the $source_field_config array to pass this information, as you did before?

File: src/ConfigTemplateTypeInterface.php [hunk 9]
```diff
@@ -38,7 +44,7 @@ public function getConfigChangesForField(string $entity_type, string $bundle, st
    * @return \Drupal\testsite_builder\ConfigTemplateMerge
    *   Returns configuration template merge for config type and bundle.
    */
-  public function getConfigChangesForBundle(string $bundle, $source_config);
+  public function getConfigChangesForBundle(string $collection_id, string $entity_type, string $bundle, $source_config);
```
Comment (dbosen) [inline src/ConfigTemplateTypeInterface.php:47]: Same problem with the `collection_id` as above.

File: src/Plugin/TestsiteBuilder/ConfigTemplateType/ViewField.php [hunk 10]
```diff
@@ -23,33 +23,54 @@ class ViewField extends Generic {
...
     $storage = $this->entityTypeManager->getStorage($entity_type);
 
     $source_field_config['id'] = $field_name;
-    $source_field_config['table'] = $storage->getTableMapping()->getFieldTableName($field_name);
     $source_field_config['field'] = $field_name;
-    $source_field_config['entity_field'] = $field_name;
     $source_field_config['label'] = "Label: {$field_name}";
 
+    $source_field_config['table'] = $storage->getTableMapping()->getFieldTableName($field_name);
+    $source_field_config['entity_field'] = $field_name;
```
Comment (dbosen) [inline src/Plugin/TestsiteBuilder/ConfigTemplateType/ViewField.php:64]: Why move these two lines?

File: src/ConfigTemplateTypeInterface.php (hunk 8 above)
Comment (dbosen) [reply src/ConfigTemplateTypeInterface.php:30]: I see now, that it is used in the search api plugins a lot more. I don't know.

**Review by dbosen — CHANGES_REQUESTED** [review]
It generally works good. some smaller issues, but i am concerned about the function signatures of the plugins. collection_id and the config array are both used only partially, maybe this can somehow be unified?

File: src/ConfigTemplateDefinition.php (hunk 1 above)
Comment (ol0lll) [reply src/ConfigTemplateDefinition.php:103]: > not all places, where this is called are handling the NULL correctly. e.g in class ConfigTemplateDefinitionCollection it is assumed, that this is always an array. also a couple of times in class ConfigTemplateDefinitionResolver.

I fixed handling NULL values by checking the return value in aforementioned places.

> Also, is it correct, that this can return string or array? This is a bit strange.

It looks like this is just returning values from [testsite_builder.config_template.views_node_content.yml](https://github.com/thunder/testsite_builder/blob/feat/refactoring-of-config-template-importer/template/testsite_builder.config_template.views_node_content.yml) that might be an array or a string.

File: src/ConfigTemplateDefinitionCollection.php (hunk 3 above)
Comment (ol0lll) [reply src/ConfigTemplateDefinitionCollection.php:88]: Well, then it should be fixed by handling null / false values in aforementioned places?

File: src/ConfigTemplateDefinition.php (hunk 2 above)
Comment (ol0lll) [reply src/ConfigTemplateDefinition.php:112]: I added checks

File: src/ConfigTemplateTypeInterface.php (hunk 8 above)
Comment (ol0lll) [reply src/ConfigTemplateTypeInterface.php:30]: Well I think we need this for the facets on search api view.

File: src/ConfigTemplateDefinition.php (hunk 6 above)
Comment (ol0lll) [reply src/ConfigTemplateDefinition.php]: removed

File: src/ConfigTemplateDefinitionResolver.php (hunk 4 above)
Comment (ol0lll) [reply src/ConfigTemplateDefinitionResolver.php:86]: okay

File: src/ConfigTemplateImporter.php (hunk 7 above)
Comment (ol0lll) [reply src/ConfigTemplateImporter.php:150]: `¯\_(ツ)_/¯`

**Review by dbosen — APPROVED** [review]
